mne/preprocessing/hfc.py

```diff
@@ -0,0 +1,135 @@
+# Authors: George O'Neill <g.o'neill@ucl.ac.uk>
+#
+# License: BSD-3-Clause
+
+import numpy as np
+from .maxwell import _prep_mf_coils, _sss_basis
+from ..forward import _prep_meg_channels
+from ..io.pick import pick_types
+from ..io.proj import Projection
+from ..utils import verbose
+
+
+@verbose
+def compute_proj_hfc(info, order=1, picks=None, exclude_bads=True,
+                     accuracy='accurate', verbose=None):
+    """Generate projectors to perform homogeneous/harmonic correction to data.
+
+    Remove evironmental fields from magentometer data by assuming it is
+    explained as a homogeneous :footcite:`TierneyEtAl2021` or harmonic field
+    :footcite:`TierneyEtAl2022`. Useful for arrays of OPMs.
+
+    Parameters
+    ----------
+    info : instance of Info
+        Info from an instance of Raw (preferable).
+    order : int
+        The order of the spherical harmonic basis set to use. Set to 1 to use
+        only the homogeneous field component (default), 2 to add gradients, 3
+        to add quadrature terms etc.
+    picks : int | array-like
+        List of channels indices to compute projeciton on. Default behaviour
+        is to select meg (non-reference) with known positions.
+    exclude_bads : bool
+        Do not include bad channels in the projection, (default: True).
+    accuracy : str
+        Can be ``"point"``, ``"normal"`` or ``"accurate"`` (default), defines
+        which level of coil definition accuracy is used to generate model.
+    %(verbose)s
+
+    Returns
+    -------
+    projs: list
+        List of projection vectors.
+
+    Notes
+    -----
+    To apply the projectors to a dataset, use
+    ``[DATA].add_proj(projs).apply_proj()``.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+    if picks is None:
+        if exclude_bads is True:
+            picks = pick_types(info, meg=True)
+        else:
+            picks = pick_types(info, meg=True, exclude=None)
+    picks = _filter_channels_with_positions(info, picks)
+    basis, channels = _generate_basis_set(info, picks, order, accuracy)
+    labels = _label_basis(order)
+    projs = []
+    for ii in range(len(labels)):
+        data = basis[:, ii]
+        proj_data = dict(col_names=channels, row_names=None,
+                         data=data[np.newaxis, :], ncol=len(channels), nrow=1)
+        proj = Projection(active=False, data=proj_data, desc=labels[ii])
+        projs.append(proj)
+
+    return projs
+
+
+def _filter_channels_with_positions(info, indsin):
+    """Keep indices of channels with position information."""
+    ch_inds = list()
+    for ii in indsin:
+        if not (any(np.isnan(info['chs'][ii]['loc']))):
+            ch_inds.append(ii)
+    return ch_inds
+
+
+def _generate_basis_set(info, picks=None, order=1,
+                        accuracy='accurate', origin=(0, 0, 0)):
+    """Generate the basis set used for HFC."""
+    exp = dict(origin=(0, 0, 0), int_order=0, ext_order=order)
```

Comment:
I think we should actually expose `origin` and `coord_frame` as `kwargs` to the `compute_projs_hfc`. IIRC they will matter for higher expansion orders, and the device origin isn't always a great choice. For consistency it's probably best to make the defaults `origin='auto', coord_frame='head'`

Reply:
I understand what this is important for the irregular spherical harmonics used for SSS's inner space. But my understanding is the span of the basis set is invariant to origin or coordinate frame used for the regular spherical harmonics used for the external basis set?

Reply:
I like this game.

```python
raw = mne.io.read_raw_fif(mne.datasets.sample.data_path() / 'MEG' / 'sample' / 'sample_audvis_raw.fif')
idx = mne.pick_types(raw.info,meg=True)

basis1, names1 = mne.preprocessing.hfc._generate_basis_set(raw.info,picks=idx,order=3,origin=(0,0,0))
basis2, names2 = mne.preprocessing.hfc._generate_basis_set(raw.info,picks=idx,order=3,origin=(0.02, 0.03, 0.04))

plt.plot(basis1[:,-1])
plt.plot(basis2[:,-1],'--')
```

![image](https://user-images.githubusercontent.com/3579812/232106545-70ca32cb-05bb-4e80-909b-84501554d370.png)

Reply:
I am confused but also convinced by this :)

Reply:
No hang on I have found a bug. Hello Hubris.

Reply:
Wait no I'm not. You don't pass origin
```suggestion
    exp = dict(origin=origin, int_order=0, ext_order=order)
```

Reply:
Yes I have no idea how the slipped the net, I can replicate your finding.

https://github.com/mne-tools/mne-python/pull/11536#discussion_r1166076414

Reply:
Yes I have no idea how that slipped the net. I can replicate you now. 
![image](https://user-images.githubusercontent.com/3579812/232108063-15b7ead2-9cc9-43c3-85c1-a9276787a4ef.png)

Reply:
Nice! So how about adding as arguments but use default coord frame 'meg' and origin at zero? Assuming OPM manufacturers set a same device origin it's a good reason to make the default different from Maxwell filter where things need to work for a larger helmet with a mobile head inside it somewhere

Reply:
Hi guys, 

@georgeoneill is correct here the origin argument is not relevant. The basis vectors themselves may be different depending on the origin but their span is not. The SPM code that verifies this is below (up to order 8 as an example)

```
%non spherical surfaces 
sp = gifti([spm('dir'),'\canonical\scalp_2562.surf.gii']);
nv = spm_mesh_normals(sp);

% general solution
S=[];
S.li=8;
S.reg=1;
S.v=double(sp.vertices);
S.o=nv;
S.scale=1;
H1= spm_opm_vslm(S);

% residuals with first basis set
beta = randn(size(H1,2),1);
test = H1*beta;
mod1 = H1*(pinv(H1)*test);
resid = test-mod1;
var(resid)/var(test)


% new origin 
v = double(sp.vertices);
v(:,1)= v(:,1)+15;
v(:,2)= v(:,2)-55;
v(:,3)= v(:,3)+65;

S=[];
S.li=8;
S.reg=1;
S.v=v;
S.o=nv;
S.scale=1;
H2= spm_opm_vslm(S);

% residuals of new basis set 
mod2 = H2*(pinv(H2)*test);
resid = test-mod2;
var(resid)/var(test)

 %mathetmatical conditions for matrices A and B spanning same space: B=AC, A=B*inv(C)

 A = H1;
 B = H2;
 C = pinv(H1)*H2;
 
 mean(mean(abs(B-A*C)))
 mean(mean(abs(A-B*inv(C))))
`

Reply:
Nice! Subspace angle was the next thing I was going to check, but I'm sure that would have been redundant with this. Okay for omitting the origin argument then.